### ISSUE TITLE
Rename a constructor renames all record fields

### ISSUE THREAD

**Issue by deemp** [haskell/haskell-language-server#2915]
### Your environment

Which OS do you use:
Ubuntu 20.04
Which LSP client (editor/plugin) do you use:
VS Code, Haskell extension v2.2.0, ghc 9.2.0, HLS 1.7.0.0
Describe your project (alternative: link to the project):
A .yml parser.  

### Steps to reproduce

See examples in https://github.com/haskell/haskell-language-server/issues/3106#issue-1338974881

### Expected behaviour

Only the constructor name is updated

### Actual behaviour

All fields of a record are renamed the same as the new consructor name

**Comment by OliverMadine** [comment]
Hi, 

Thanks for the report!

Please could you update the link to the commit for the steps to reproduce? It looks like the repository might be private.

@br4ch1st0chr0n3

**Comment by br4ch1st0chr0n3** [comment]
Hi, sorry for that. Should be public now, @OliverMadine.

**Comment by OliverMadine** [comment]
It seems like the issue is hie files including record field names as references to the constructor.
This is only happening with ghc > 9.

@wz1000 maybe you have some more insight?

**Comment by pepeiborra** [comment]
> It seems like the issue is hie files including record field names as references to the constructor.
> This is only happening with ghc >= 9.

Is this a bug or a feature? 

If it is a bug, we will need to report it in the GHC bug tracker and ideally contribute a fix. And back port the fixes to hie-compat: https://github.com/haskell/haskell-language-server/tree/master/hie-compat

**Comment by michaelpj** [comment]
Is this still a problem?

**Comment by konn** [comment]
Yes. I've gotten bitten by this yesterday at least with HLS 2.5.0.0.

**Comment by konn** [comment]
I've just tested and confirmed that it still occurs with HLS 2.6.0.0.

**Comment by soulomoon** [comment]
Perhaps the `type RefMap a = Map Identifier [(Span, IdentifierDetails a)]` should contains the [NodeOrigin](https://hackage.haskell.org/package/ghc-9.8.2/docs/src/GHC.Iface.Ext.Types.html#NodeOrigin) to indicate if it is one of the evidence variable

**Comment by jhrcek** [comment]
Note that I took a stab at fixing this in https://github.com/haskell/haskell-language-server/pull/4089

That fix works, but requires changing on hiedb side (skipping indexing of all Identifiers whose NodeOrigin is GeneratedInfo).
We agreed with @michaelpj that instead of that approach, we should rather add something like a boolean column for indexed identifiers saying whether that identifier is generated or not.
This would then be used by the rename plugin which would filter out all generated occurrences of given identifier, thus avoiding this bug via different route.

**Comment by soulomoon** [comment]
I also make an issue for the type change as well https://gitlab.haskell.org/ghc/ghc/-/issues/24751

**Comment by soulomoon** [comment]
> Is https://github.com/wz1000/HieDb/pull/77 sufficient to solve this? Or do we still need a solution on GHC-side?

Yes, I think so

**Comment by SmartHypercube** [comment]
> > Is [wz1000/HieDb#77](https://github.com/wz1000/HieDb/pull/77) sufficient to solve this? Or do we still need a solution on GHC-side?
> 
> Yes, I think so

(English is not my native language, please excuse me if the answer is obvious.) Does this reply mean "Yes, I think ... is sufficient to solve this" or "Yes, I think we still need a solution on GHC-side"?

**Comment by soulomoon** [comment]
@SmartHypercube . Former, but better if we do it in ghc too.

**Comment by deemp** [comment]
Thanks a lot! Didn't yet try though.
